modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/ContinuousQueryBufferCleanupAbstractTest.java

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Test for continuous query buffer cleanup.
+ */
+public abstract class ContinuousQueryBufferCleanupAbstractTest extends GridCommonAbstractTest {
+    /** */
+    private static final int RECORDS_CNT = 10000;
+
+    /** */
+    private static final int ACK_THRESHOLD = 100;
+
+    /** */
+    private static final String REMOTE_ROUTINE_INFO_CLASS_NAME = "org.apache.ignite.internal.processors.continuous.GridContinuousProcessor$RemoteRoutineInfo";
+
+    /** */
+    private static final String LOCAL_ROUTINE_INFO_CLASS_NAME = "org.apache.ignite.internal.processors.continuous.GridContinuousProcessor$LocalRoutineInfo";
+
+    /** */
+    private static final String BATCH_CLASS_NAME = "org.apache.ignite.internal.processors.cache.query.continuous.CacheContinuousQueryEventBuffer$Batch";
+
+    /** Continuous query updates counter */
+    protected final AtomicInteger updCntr = new AtomicInteger();
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * @throws Exception If fail.
+     */
+    @Test
+    public void checkCacheWithSingleNode() throws Exception {
+        checkBuffer(1, 0, true);
+    }
+
+    /**
+     * @throws Exception If fail.
+     */
+    @Test
+    public void checkCacheWithMultipleNodes() throws Exception {
+        checkBuffer(2, 0, true);
+    }
+
+    /**
+     * @throws Exception If fail.
+     */
+    @Test
+    public void checkCacheWithMultipleNodesWithBackups() throws Exception {
+        checkBuffer(2, 1, true);
+    }
+
+    /**
+     * @throws Exception If fail.
+     */
+    @Test
+    public void checkCacheWithMultipleNodesWithBackupsWithoutClient() throws Exception {
+        checkBuffer(2, 1, false);
+    }
+
+    /** Provides continuous query for the test. */
+    protected abstract AbstractContinuousQuery<Integer, String> getContinuousQuery();
+
+    /** */
+    private void checkBuffer(int srvCnt, int backupsCnt, boolean useClient) throws Exception {
+        System.setProperty("IGNITE_CONTINUOUS_QUERY_ACK_THRESHOLD", Integer.toString(ACK_THRESHOLD));
+
+        IgniteEx[] srvs = new IgniteEx[srvCnt];
+
+        for (int i = 0; i < srvCnt; i++)
+            srvs[i] = startGrid("srv" + i);
+
+        Ignite qryNode = useClient
+            ? startGrid(optimize(getConfiguration("client").setClientMode(true)))
+            : srvs[0];
+
+        CacheConfiguration<Integer, String> cacheCfg = new CacheConfiguration<Integer, String>("testCache")
+            .setBackups(backupsCnt)
+            .setAffinity(new RendezvousAffinityFunction(32, null));
+
+        IgniteCache<Integer, String> cache = qryNode.getOrCreateCache(cacheCfg);
+
+        AbstractContinuousQuery<Integer, String> qry = getContinuousQuery();
+
+        cache.query(qry);
+
+        for (int i = 0; i < RECORDS_CNT; i++)
+            cache.put(i, Integer.toString(i));
+
+        Thread.sleep(2000);
```

Comment:
Why do we need `sleep` here? I think we should use the `waitForCondition` here that all of the entries are processed by CQ.

Reply:
waitForCondition used instead of sleep.